go/oasis-test-runner/cmd/cmp.go

```diff
@@ -0,0 +1,517 @@
+	"github.com/spf13/cobra"
+	flag "github.com/spf13/pflag"
+	"github.com/spf13/viper"
+
+	"github.com/oasislabs/oasis-core/go/common/logging"
+	"github.com/oasislabs/oasis-core/go/oasis-node/cmd/common"
+	"github.com/oasislabs/oasis-core/go/oasis-node/cmd/common/metrics"
+	testOasis "github.com/oasislabs/oasis-core/go/oasis-test-runner/oasis"
+)
+
+const (
+	cfgMetrics                = "metrics"
+	cfgMetricsP               = "m"
+	cfgMetricsTargetGitBranch = "metrics.target.git_branch"
+	cfgMetricsSourceGitBranch = "metrics.source.git_branch"
+	cfgMetricsNetDevice       = "metrics.net.device"
+)
+
+var (
+	cmpCmd = &cobra.Command{
+		Use:   "cmp",
+		Short: "compare last two benchmark instances",
+		Long: `cmp connects to prometheus, fetches the metrics of the benchmark instances and
+compares them. By default, the most recent instance (source) is fetched and
+compared to the pre-last (target). If --metrics.{target|source}.git_branch is
+provided, it compares the most recent instances in the corresponding branches.
+cmp compares all metrics provided by --metrics parameter and computes ratio
+source/target of metric values. If any of the metrics exceeds
+max_threshold.<metric>.{avg|max}_ratio or doesn't reach
+min_threshold.<metric>.{avg|max}_ratio, ba exits with error code 1.`,
+		Run: runCmp,
+	}
+
+	allMetrics = map[string]*Metric{
+		"time": &Metric{
+			getter:               getDuration,
+			maxThresholdAvgRatio: 1.1,
+			maxThresholdMaxRatio: 1.1,
+		},
+		"du": &Metric{
+			getter:               getDiskUsage,
+			maxThresholdAvgRatio: 1.06,
+			maxThresholdMaxRatio: 1.15,
+		},
+		"io": &Metric{
+			getter:               getIOWork,
+			maxThresholdAvgRatio: 1.2,
+			maxThresholdMaxRatio: 1.2,
+		},
+		"mem": &Metric{
+			getter:               getRssAnonMemory,
+			maxThresholdAvgRatio: 1.1,
+			maxThresholdMaxRatio: 1.1,
+		},
+		"cpu": &Metric{
+			getter:               getCPUTime,
+			maxThresholdAvgRatio: 1.05,
+			maxThresholdMaxRatio: 1.05,
+		},
+		"net": &Metric{
+			getter: getNetwork,
+			// Network stats suffer effects from other processes too and varies.
+			maxThresholdAvgRatio: 1.3,
+			maxThresholdMaxRatio: 1.3,
+		},
+	}
+	userMetrics []string
+
+	client api.Client
+
+	cmpLogger *logging.Logger
+)
+
+type Metric struct {
+	getter               func(context.Context, string, *model.SampleStream) (float64, float64, error)
+	maxThresholdAvgRatio float64
+	maxThresholdMaxRatio float64
+	minThresholdAvgRatio float64
+	minThresholdMaxRatio float64
+}
+
+// getDuration returns average and maximum running times of the given coarse benchmark instance ("up" metric w/ minute
+// resolution time series).
+func getDuration(ctx context.Context, test string, bi *model.SampleStream) (float64, float64, error) {
+	instance := string(bi.Metric[testOasis.MetricsLabelInstance])
+
+	// Re-fetch the given benchmark instance with second resolution. Each obtained time series corresponds to one run.
+	v1api := v1.NewAPI(client)
```

Comment:
Ugh, this `v1` is so confusing, can you rename the package import to something else, e.g. `prometheusAPIV1`.

Reply:
I just use `prometheusAPI`. `V1` suffix seems overkill for our use.

Reply:
Yeah that sounds reasonable.